### PR TITLE
feat(attention): FA3 kernels for a 16-bit query with an FP8 KV cache

### PR DESCRIPTION
**Pull request by SamMausberg**
## Description

FA3 (SM90) attention for a 16-bit query (`bfloat16` / `float16`) with an FP8 (`e4m3` / `e5m2`) KV cache, for `single_prefill_with_kv_cache`, `BatchPrefillWithRaggedKVCacheWrapper` and `BatchPrefillWithPagedKVCacheWrapper`. Until now FA3 rejected this dtype pair, so `backend="auto"` on H100/H200 fell back to the FA2 mixed-precision kernel, which is slower than FA2 with a 16-bit cache.

**Kernel.** The MMA warpgroups run the unchanged 16-bit consumer. Only the producer side is new (`hopper/dequant_mainloop.cuh`): K/V tiles are loaded as FP8 into small staging buffers (TMA for dense K/V, the existing cp.async page-table gather for paged K/V), and all 128 producer threads dequantize each staged tile into the 16-bit swizzled K/V layout the consumer already reads. The conversion is exact, so the kernel computes the same values as the FA2 mixed-precision kernel and is bit-identical to the 16-bit FA3 kernel on the dequantized cache.

Tile shapes are those of the 16-bit kernels, except that dense head_dim 128 keeps `CTA_KV = 128` for the non-causal kernel and head_dim 256 keeps one staging stage so the staging buffers fit in shared memory. ptxas reports no spills.

**Python.** `backend="auto"` selects fa3 for this dtype pair on SM90 for prefill (also through the block-sparse and attention-sink wrappers). Tensor-core decode keeps fa2, whose short query tiles are 4-5x faster than fa3 for decode. `k_scale` / `v_scale` keep their semantics (the ragged wrapper's pre-existing gap is #4979). `aot.py` generates the new modules.

**Performance.** GH200 480GB, CUDA 12.8, `benchmarks/flashinfer_benchmark.py`, bf16 query, e4m3 KV cache, 32/8 heads, head_dim 128, page size 16, median of 50 iterations. `fa2` is what `backend="auto"` selected before this PR; the last column is the 16-bit fa3 kernel on the same shapes.

| kernel | mask | batch x qo_len / kv_len | fa2, fp8 KV (before) | fa3, fp8 KV (this PR) | speedup | fa3, bf16 KV |
|---|---|---|---|---|---|---|
| paged | causal | 4 x 4096 / 4096 | 2.195 ms | 1.699 ms | 1.29x | 1.242 ms |
| paged | causal | 8 x 1024 / 8192 | 3.767 ms | 2.912 ms | 1.29x | 2.174 ms |
| paged | causal | 1 x 16384 / 16384 | 8.120 ms | 5.974 ms | 1.36x | 4.402 ms |
| paged | non-causal | 4 x 4096 / 4096 | 4.007 ms | 3.120 ms | 1.28x | 2.267 ms |
| paged | non-causal | 8 x 1024 / 8192 | 3.954 ms | 3.065 ms | 1.29x | 2.260 ms |
| paged | non-causal | 1 x 16384 / 16384 | 15.549 ms | 11.900 ms | 1.31x | 8.649 ms |
| ragged | causal | 4 x 4096 / 4096 | 2.008 ms | 1.234 ms | 1.63x | 0.842 ms |
| ragged | causal | 8 x 1024 / 8192 | 3.547 ms | 2.157 ms | 1.64x | 1.438 ms |
| ragged | causal | 1 x 16384 / 16384 | 7.551 ms | 4.473 ms | 1.69x | 2.950 ms |
| ragged | non-causal | 4 x 4096 / 4096 | 3.702 ms | 2.295 ms | 1.61x | 1.568 ms |
| ragged | non-causal | 8 x 1024 / 8192 | 3.676 ms | 2.254 ms | 1.63x | 1.483 ms |
| ragged | non-causal | 1 x 16384 / 16384 | 14.414 ms | 8.906 ms | 1.62x | 6.121 ms |

FA2 with a bf16 cache takes 1.751 ms (paged) and 1.554 ms (ragged) on the first row's shape, so the fp8 cache is now faster than the 16-bit FA2 path as well. The remaining gap to the 16-bit fa3 kernel is the dequantization competing with the consumer warps for issue slots.

**Prior art.** The same staging-plus-dequantization idea is used by the SM90 FP8-KV kernel @qixiang-99 added to vLLM's flash-attention fork (vllm-project/flash-attention#147, CuTe DSL, dequantization in the consumer warpgroups); thanks to @jhaotingc for the pointer. This is an independent implementation for FlashInfer's C++ FA3 kernels that also covers the paged and single-prefill paths.

## Related Issues

Closes the FlashInfer side of #3327. Resolves the SM90 part of #1753. Found on the way and filed separately: #4976 (fa2 with an fp8 cache is wrong for the (192, 128) head-dim pair), #4979 (the ragged wrapper ignores `k_scale` / `v_scale` for an fp8 cache) and a wider scope for #4972 (fa2 non-causal sliding window).

## Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

New file `tests/attention/test_hopper_fp8_kv.py`, 690 cases, all passing on GH200: single, ragged and paged prefill for every `{bf16, fp16} x {e4m3, e5m2}` pair at head_dim 128 (each pair is a different conversion path) and at head_dim 64 / 256, page sizes 1 / 5 / 16, NHD and HND, chunked prefill, causal and non-causal, sliding window, logits soft cap, `k_scale` / `v_scale`, the (192, 128) head-dim pair, multi-item scoring, the block-sparse and attention-sink wrappers, `backend="auto"` routing, and tensor-core decode through fa3. Every case is checked against an fp32 reference over the dequantized cache and, where fa2 is correct, against fa2. The 16-bit and fp8-query subsets of `tests/attention/test_hopper.py` and `test_hopper_fp8_attention.py` still pass.

## Experimental Track

Not applicable.

## Reviewer Notes

- `DequantKVSchedule::run` publishes tiles in the order the consumer waits for them and keeps each staging ring `NUM_STAGES_KV_STAGING` tiles ahead; the comments there give the ordering argument, including why the V staging loads no longer wait for `barrier_O` while the 16-bit V writes still do.
- `benchmarks/flashinfer_benchmark.py --refcheck` compares fa2 against fa3. With the benchmark's fp8 data (values scaled to the fp8 range, near one-hot softmax) the two disagree on a few elements in 67M; the unmodified 16-bit fa2 and fa3 kernels disagree on exactly the same elements when fed the dequantized values, so this is accumulation-order sensitivity of that data, not a kernel difference.
- Pre-existing, unchanged: multi-item scoring with `prefix_len == 0` and a non-empty outside-items window makes the consumer wait for KV tile -1 (the old paged loader reads the page table out of bounds there; the new one does not stage that tile).

Developed in combination with Claude Fable 5.1.

## Summary by CodeRabbit

- **New Features**
  - Added support for Hopper FA3 attention with FP16 or BF16 queries and FP8 key-value caches.
  - Extended support across prefill modes, paged and ragged batches, sliding windows, soft caps, calibration scales, block-sparse attention, attention sinks, and multi-item scoring.
  - Automatic backend selection now routes supported FP8-cache workloads appropriately.

- **Bug Fixes**
  - Corrected backend selection for FP8 KV caches when queries use non-FP8 16-bit dtypes.

### CODE REVIEW

File: flashinfer/utils.py [hunk 1]
```diff
@@ -468,10 +468,13 @@ def is_fa3_backend_supported(
...
     if use_fp16_qk_reductions:
         return False
-    # FA3 FP8 KV cache currently requires FP8 query.
+    # FA3 supports an FP8 KV cache with either an FP8 query (FP8 tensor cores) or a
+    # 16-bit query (K/V dequantized to the query dtype in the kernel).
     if dtype_kv in {torch.float8_e4m3fn, torch.float8_e5m2} and dtype_q not in {
         torch.float8_e4m3fn,
         torch.float8_e5m2,
+        torch.float16,
+        torch.bfloat16,
```
[the remark this answers is not in the document]
Comment (SamMausberg) [reply flashinfer/utils.py:477]: Checked: this is the same situation as the existing 16-bit fa3 path, the new dtype pair does not add. `o_dtype` is independent of the KV cache dtype for the fa3 templates; the `gen_fa3` condition only limits which modules `aot.py` prebuilds (16-bit query modules are prebuilt with `dtype_o == dtype_q`), and JIT builds the others on demand, exactly as it already does for a 16-bit query with a 16-bit cache. Verified on a GH200: `single_prefill_with_kv_cache(q=bf16, k=v=e4m3, o_dtype=float16, backend="fa3")` builds and matches the bf16-output result to 0.008. No change.

**Comment by SamMausberg** [comment]
Follow-up commit: the `THREADS_PER_GROUP <= 32` static_assert from the review is added, along with static_asserts for the shared-memory budget and the 8-row swizzle atoms the dequantizer relies on, a multi-item-scoring test shape that exercises the tile jump, and a bit-exact comparison of the paged kernel against the 16-bit fa3 kernel on the dequantized cache.

**Comment by SamMausberg** [comment]
One last thing worth considering is adapting the page-aware TMA loading from #4971 to the FP8 staging buffers here, while keeping dequantization into the existing 16-bit layout.

It could be worth benchmarking against the current paged cp.async path, though the gain may be limited if dequantization dominates.
